kolibri/content/utils/channel_import.py

```diff
@@ -376,33 +378,57 @@ def check_and_delete_existing_channel(self):
         except ChannelMetadata.DoesNotExist:
             existing_channel = None
+
         if existing_channel:
 
             if existing_channel.version < self.channel_version:
                 # We have an older version of this channel, so let's clean out the old stuff first
                 logging.info(('Older version {channel_version} of channel {channel_id} already exists in database; removing old entries ' +
                               'so we can upgrade to version {new_channel_version}').format(
                     channel_version=existing_channel.version, channel_id=self.channel_id, new_channel_version=self.channel_version))
+                self.delete_old_channel_data(existing_channel.root.tree_id)
             else:
                 # We have previously loaded this channel, with the same or newer version, so our work here is done
                 logging.warn(('Version {channel_version} of channel {channel_id} already exists in database; cancelling import of ' +
                               'version {new_channel_version}').format(
                     channel_version=existing_channel.version, channel_id=self.channel_id, new_channel_version=self.channel_version))
                 return False
-            self.check_cancelled()
-            root_id = ChannelMetadata.objects.get(pk=self.channel_id).root_id
-            ContentNodeClass = self.destination.get_class(ContentNode)
-            root_node = self.destination.session.query(ContentNodeClass).get(root_id)
-            # We set cascade behaviour on relationships on the default database
-            # So calling delete on the root node of a channel will delete everything else
-            self.destination.session.delete(root_node)
-            # Deletion/cascade is not enacted on the session until we call flush
-            # Note: as this function is generally called inside a explicitly managed
-            # transaction, this will not be committed to the database unless the entire
-            # update operation succeeds.
-            self.destination.session.flush()
+
         return True
 
+    def _can_use_optimized_pre_deletion(self):
+        # check whether we can skip fully deleting these three models, if we'll be using REPLACE on them anyway
+        for model in [ContentNode, File, AssessmentMetaData]:
+            mapping = self.schema_mapping.get(model, {})
+            row_mapper = self.generate_row_mapper(mapping.get('per_row'))
+            table_mapper = self.generate_table_mapper(mapping.get('per_table'))
+            if not self.can_use_sqlite_attach_method(model, row_mapper, table_mapper):
+                return False
+        return True
+
+    def delete_old_channel_data(self, old_tree_id):
+
+        # construct a template for deleting records for models that have a foreign key onto ContentNode
+        delete_related_template = "DELETE FROM {table} WHERE contentnode_id IN (SELECT id FROM content_contentnode WHERE tree_id = '{tree_id}')"
```

Comment:
Would be nice if we could avoid hardcoding the SQL table names and column names - if we can derive them from the SQLAlchemy table definitions would make this feel a bit DRYer.